backend/libinput/tablet_tool.c

```diff
@@ -1,25 +1,179 @@
-	struct wlr_tablet_tool *wlr_tablet_tool = calloc(1, sizeof(struct wlr_tablet_tool));
-	if (!wlr_tablet_tool) {
+	struct wlr_libinput_tablet *libinput_tablet_tool =
+		calloc(1, sizeof(struct wlr_libinput_tablet));
+	if (!libinput_tablet_tool) {
 		wlr_log(L_ERROR, "Unable to allocate wlr_tablet_tool");
 		return NULL;
 	}
-	wlr_tablet_tool_init(wlr_tablet_tool, NULL);
+	struct wlr_tablet_tool *wlr_tablet_tool = &libinput_tablet_tool->wlr_tool;
+
+	wlr_list_init(&wlr_tablet_tool->paths);
+	struct udev_device *udev = libinput_device_get_udev_device(libinput_dev);
+	wlr_list_push(&wlr_tablet_tool->paths, strdup(udev_device_get_syspath(udev)));
+	wlr_tablet_tool->name = strdup(libinput_device_get_name(libinput_dev));
+	wl_list_init(&libinput_tablet_tool->tools);
+
+	wlr_tablet_tool_init(wlr_tablet_tool, &tool_impl);
 	return wlr_tablet_tool;
 }
 
+static enum wlr_tablet_tool_type wlr_type_from_libinput_type(
+		enum libinput_tablet_tool_type value) {
+	switch (value) {
+	case LIBINPUT_TABLET_TOOL_TYPE_PEN:
+		return WLR_TABLET_TOOL_TYPE_PEN;
+	case LIBINPUT_TABLET_TOOL_TYPE_ERASER:
+		return WLR_TABLET_TOOL_TYPE_ERASER;
+	case LIBINPUT_TABLET_TOOL_TYPE_BRUSH:
+		return WLR_TABLET_TOOL_TYPE_BRUSH;
+	case LIBINPUT_TABLET_TOOL_TYPE_PENCIL:
+		return WLR_TABLET_TOOL_TYPE_PENCIL;
+	case LIBINPUT_TABLET_TOOL_TYPE_AIRBRUSH:
+		return WLR_TABLET_TOOL_TYPE_AIRBRUSH;
+	case LIBINPUT_TABLET_TOOL_TYPE_MOUSE:
+		return WLR_TABLET_TOOL_TYPE_MOUSE;
+	case LIBINPUT_TABLET_TOOL_TYPE_LENS:
+		return WLR_TABLET_TOOL_TYPE_LENS;
+	}
+
+	assert(false && "UNREACHABLE");
+}
+
+static struct wlr_libinput_tablet_tool *get_wlr_tablet_tool(
+		struct libinput_tablet_tool *tool) {
+	struct wlr_libinput_tablet_tool *ret =
+		libinput_tablet_tool_get_user_data(tool);
+
+	if (ret) {
+		return ret;
+	}
+
+	ret = calloc(1, sizeof(struct wlr_libinput_tablet_tool));
+	if (!ret) {
+		return NULL;
+	}
+
+	ret->libinput_tool = libinput_tablet_tool_ref(tool);
+	ret->wlr_tool.pressure = libinput_tablet_tool_has_pressure(tool);
+	ret->wlr_tool.distance = libinput_tablet_tool_has_distance(tool);
+	ret->wlr_tool.tilt = libinput_tablet_tool_has_tilt(tool);
+	ret->wlr_tool.rotation = libinput_tablet_tool_has_rotation(tool);
+	ret->wlr_tool.slider = libinput_tablet_tool_has_slider(tool);
+	ret->wlr_tool.wheel = libinput_tablet_tool_has_wheel(tool);
+
+	ret->wlr_tool.hardware_serial = libinput_tablet_tool_get_serial(tool);
+	ret->wlr_tool.hardware_wacom = libinput_tablet_tool_get_tool_id(tool);
+	ret->wlr_tool.type = wlr_type_from_libinput_type(
+		libinput_tablet_tool_get_type(tool));
+
+	ret->unique = libinput_tablet_tool_is_unique(tool);
+
+	wl_signal_init(&ret->wlr_tool.events.destroy);
+
+	libinput_tablet_tool_set_user_data(tool, ret);
+	return ret;
+}
+
+static void ensure_tool_reference(struct wlr_libinput_tablet_tool *tool,
+		struct wlr_tablet_tool *wlr_dev) {
+	struct tablet_tool_list_elem *pos;
+	struct wlr_libinput_tablet *tablet = wl_container_of(wlr_dev, tablet, wlr_tool);
```

Comment:
Is it possible to add assertions to all of these `wl_container_of`?

Reply:
What kind of assertions? Assert non-NULL, or should we have a better way to check for valid arguments?

Reply:
No need to assert non-NULL. I was thinking about checking that the `wlr_dev` comes from the libinput backend.